doc: Adds example for `mongodbatlas_backup_compliance_policy` (#3391)
mongodb/terraform-provider-mongodbatlas: examples/mongodbatlas_backup_compliance_policy/module/README.md

# Example - MongoDB Atlas Backup Compliance Policy with a module

> This example is identical to the [resource example](../resource/README.md) except that it uses a local [module](modules/cluster_with_schedule/main.tf) to manage the `mongodbatlas_advanced_cluster` and the `mongodbatlas_cloud_backup_schedule` via the `add_schedule` variable.
> 
> The local module is meant as a placeholder for an enterprise module that cannot be changed by the module user and the module user doesn't have access to run `terraform state` commands.
> 
> The [cleanup step below](#different-cleanup-when-using-the-removed-block-for-a-module) shows how the `moved` and `removed` block can remove the `mongodbatlas_cloud_backup_schedule` from your Terraform state to avoid the [BACKUP_POLICIES_NOT_MEETING_BACKUP_COMPLIANCE_POLICY_REQUIREMENTS](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled) error when running `terraform destroy`.

Reviewer: ```suggestion
This example module is identical to the [resource example](../resource/README.md) except that it is designed to serve as a reference for platform teams who have created their own module and make it available to internal teams for leveraging the MongoDB Atlas Terraform provider. Typically, these users do not have the ability to execute `terraform state` commands or modify the Terraform state manually.

Similar to the resource example, the emphasis here is on how to prevent the [BACKUP_POLICIES_NOT_MEETING_BACKUP_COMPLIANCE_POLICY_REQUIREMENTS](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled) error when running `terraform destroy` on a `mongodbatlas_advanced_cluster` that includes a `mongodbatlas_cloud_backup_schedule` and has a backup compliance policy enabled.  

To achieve this, we'll leverage:  
- An `add_schedule` variable, provided by the module maintainer, to control whether the `mongodbatlas_cloud_backup_schedule` is included in the configuration.  
- The `moved` and `removed` blocks, implemented at the root level by the module user (see more details about the `removed` block limitation in the Hashicorp Terraform [issue 34439](https://github.com/hashicorp/terraform/issues/34439))
```

> ## Different cleanup when using the `removed` block for a module

Reviewer: ```suggestion
## How to delete the cluster and retain their backup snapshots
```

the title is hard to understand. Remember a customer doesn't want to use `removed`, they just want to **delete a cluster retaining their backups**

Author: Thanks changed!

> **Note**: If you can use the `terraform state rm` command or edit the module TF files directly, follow the simpler steps in the [resource example readme](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled).

Reviewer: ```suggestion
```

suggest to remove this at this point. audience of the example is already explained above.

> If you try to use the `removed` block without deleting the `from` resource you get the error: `Removed Resource still exists error`:

Reviewer: I find this section redundant, and also kind of breaking the flow of the text because you're saying that if they do X (which we're not telling them to do) it won't work.

If you really want, this should go at the very bottom in a sort of FAQ

```suggestion
```

> ```terraform
> removed {
>   from = module.cluster_with_schedule.mongodbatlas_cloud_backup_schedule.this
> 
>   lifecycle {
>     destroy = false
>   }
> }
> ```

Reviewer: ```suggestion

```

> To workaround this limitation (see more details about the `removed` block limitation in the Hashicorp Terraform [issue 34439](https://github.com/hashicorp/terraform/issues/34439)) you can use a `moved` block and change the `add_schedule` flag in the root [main.tf](main.tf):

Reviewer: ```suggestion
```

you can remove this part. I mentioned the issue in the intro part

Author: Thanks changed!

> 

Reviewer: ```suggestion
Let's begin by assuming you've created a module instance with the following configuration:

```terraform
module "cluster_with_schedule" {
    source = "./modules/cluster_with_schedule"
    project_id = var.project_id
    instance_size = var.instance_size
    cluster_name = var.cluster_name
    add_schedule = true
}
>> ``` 

Note: The `add_schedule` field is set to `true`, indicating that a `mongodbatlas_cloud_backup_schedule` resource has been defined, as reflected in the module's source code.

To proceed with the deletion, we'll update the configuration as follows:
```

Reviewer: @EspenAlbert please remove the >> before ```, the suggestion wasn't working

Author: Thanks changed!

```terraform

> module "cluster_without_schedule" {

Reviewer: ```suggestion
module "cluster_with_schedule" {
```

Author: Thanks changed!

    source = "./modules/cluster_with_schedule"

    project_id = var.project_id
    instance_size = var.instance_size
    cluster_name = var.cluster_name
    add_schedule = false # changed flag
}
# Rename the resource to avoid the `Removed Resource still exists error`
moved {
  from = module.cluster_with_schedule.mongodbatlas_cloud_backup_schedule.this[0] # must be deleted with the `add_schedule` variable set to false

>   to  = mongodbatlas_cloud_backup_schedule.deleted # any resource name that doesn't exist works!

Reviewer: ```suggestion
  to  = mongodbatlas_cloud_backup_schedule.to_be_deleted # any resource name that doesn't exist works!
```

Author: Thanks changed!

}

removed {

>   from = mongodbatlas_cloud_backup_schedule.deleted # any resource name that doesn't exist works!

Reviewer: ```suggestion
  from = mongodbatlas_cloud_backup_schedule.to_be_deleted # any resource name that doesn't exist works!
```

Author: Thanks changed!

  lifecycle {
    destroy = false
  }
}
```

Then when you run `terraform apply`, you should see:

```bash
[...]

> mongodbatlas_cloud_backup_schedule.deleted will no longer be managed by Terraform, but will not be destroyed

Reviewer: ```suggestion
mongodbatlas_cloud_backup_schedule.to_be_deleted will no longer be managed by Terraform, but will not be destroyed
```

Author: Thanks changed!

 (destroy = false is set in the configuration)
  (moved from module.cluster_with_schedule.mongodbatlas_cloud_backup_schedule.this[0])
[...]
Plan: 0 to add, 0 to change, 0 to destroy.
```

Reply `yes` to confirm the state removal of `mongodbatlas_cloud_backup_schedule`.

Then run `terraform destroy` to destroy the:
- Cluster defined in `module.cluster_with_schedule.mongodbatlas_advanced_cluster.this`.

> - Root compliance policy defined in `mongodbatlas_backup_compliance_policy`.

Reviewer: can you double check this? I don't think you can delete the compliance policy from Terraform. Maybe needs language to be adjusted.

Author: Good point. Added warnings

# Example - MongoDB Atlas Backup Compliance Policy with a module
This example is identical to the [resource example](../resource/README.md) except that it uses a local [module](modules/cluster_with_schedule/main.tf) to manage the `mongodbatlas_advanced_cluster` and the `mongodbatlas_cloud_backup_schedule`.

> It shows a workaround for deleting the `mongodbatlas_advanced_cluster` and `mongodbatlas_cloud_backup_schedule` when using Terraform Modules. See more details about the `removed` block limitation in the Hashicorp Terraform [issue 34439](https://github.com/hashicorp/terraform/issues/34439).

Reviewer: Here I would add a one liner sentence to explain that the module can dynamically define `mongodbatlas_cloud_backup_schedule` resource and it uses the `moved` block to transfer from one module instance to another

Author: Added in f4869ac281896f66764ef92d3349efe438946888 WDYT?

> 

Reviewer: ```suggestion
If you can use the `terraform state rm` command, we recommend to do that following [these steps](LINK_HERE).
```

Link should go to https://github.com/mongodb/terraform-provider-mongodbatlas/pull/3391/files#diff-98b9d0de4627cbab17d9a76f3c04bde270d2e52e3e25148d3c48feaa1588b316R70

Author: Changed in f4869ac281896f66764ef92d3349efe438946888

## Different cleanup when using the `removed` block for a module

> When using a module it is likely you cannot remove the resource block from the module config directly, `module.source` can be somewhere else; therefore, you get the `Removed Resource still exists error` when trying to use the `removed` block:

Reviewer: Can you clarify:
- why are you saying likely? In what cases they can use it?
- `module.source can be somewhere else`: what do you mean with somewhere else?

Author: Changed in f4869ac281896f66764ef92d3349efe438946888
WDYT?

Author: Updated based on doc feedback

# Example - MongoDB Atlas Backup Compliance Policy with a module

> This example module is identical to the [resource example](../resource/README.md) except that it is designed to serve as a reference for platform teams who have created their own module and make it available to internal teams for leveraging the MongoDB Atlas Terraform provider. Typically, these users do not have the ability to execute `terraform state` commands or modify the Terraform state manually.

Reviewer: ```suggestion
We recommend first to read the [resource example](../resource/README.md) as it contains all the relevant information. This example module is identical to it except that it is designed to serve as a reference for platform teams who have created their own module and make it available to internal teams for leveraging the MongoDB Atlas Terraform provider. Typically, these users do not have the ability to execute `terraform state` commands or modify the Terraform state manually.
```

Author: Thank you. Changed here: 86d253d85c7759392d264dfc3f5d92dc7f6cc393

As in the resource example, the attention is focused on how to avoid the [BACKUP_POLICIES_NOT_MEETING_BACKUP_COMPLIANCE_POLICY_REQUIREMENTS](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled) when running `terraform destroy` on a `mongodbatlas_advanced_cluster` with `mongodbatlas_cloud_backup_schedule` and enabled backup compliance policy.

To do that, we'll use:
- an `add_schedule` variable provided by the module maintainer that controls the presence of the `mongodbatlas_cloud_backup_schedule` in the configuration

> - the `moved` and `removed` blocks used by the module user and added at the root level.

Reviewer: ```suggestion
- the `moved` and `removed` blocks used by the module user and added at the root level. For more info on why the two blocks are needed, please read this [Terraform issue](https://github.com/hashicorp/terraform/issues/34439)
```

sorry, I think with my changes this important link was deleted.

Author: Thank you! Added in 5b76d17165938aefc354d94e1908eb17cfd7f165

}

removed {
  from = mongodbatlas_cloud_backup_schedule.to_be_deleted # any resource name that doesn't exist works!

  lifecycle {
    destroy = false
  }
}
```

Then when you run `terraform apply`, you should see:

```bash
[...]
mongodbatlas_cloud_backup_schedule.to_be_deleted will no longer be managed by Terraform, but will not be destroyed
 (destroy = false is set in the configuration)
  (moved from module.cluster_with_schedule.mongodbatlas_cloud_backup_schedule.this[0])
[...]
Plan: 0 to add, 0 to change, 0 to destroy.
```

Reply `yes` to confirm the state removal of `mongodbatlas_cloud_backup_schedule`.

Then run `terraform destroy` to destroy the cluster defined in `module.cluster_with_schedule.mongodbatlas_advanced_cluster.this`.
See [Backup Compliance Policy Deletion Consideration](../resource/README.md#backup-compliance-policy-deletion-consideration) for details on the `mongodbatlas_backup_compliance_policy` deletion.

## FAQ
I get a `Removed Resource still exists error` when running `terraform apply`, how do I fix it?

This error happens because the configuration still has the `mongodbatlas_cloud_backup_schedule` active.

> Remember to add the `moved` block and set `add_schedule=false` on the `cluster_with_schedule` module.

Reviewer: ```suggestion
When using a module, together with the `removed` block, remember to add the `moved` block and set `add_schedule=false` on the `cluster_with_schedule` module.
```

Author: Thank you. Changed here: 86d253d85c7759392d264dfc3f5d92dc7f6cc393

# Example - MongoDB Atlas Backup Compliance Policy with a module
This example is identical to the [resource example](../resource/README.md) except that it uses a local [module](modules/cluster_with_schedule/main.tf) to manage the `mongodbatlas_advanced_cluster` and the `mongodbatlas_cloud_backup_schedule` via the `add_schedule` variable.

> The [cleanup step below](#different-cleanup-when-using-the-removed-block-for-a-module) shows how the `moved` and `removed` block can remove the `mongodbatlas_cloud_backup_schedule` from state to avoid the [BACKUP_POLICIES_NOT_MEETING_BACKUP_COMPLIANCE_POLICY_REQUIREMENTS](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled) error when running `terraform destroy`.

Reviewer 2: ```suggestion
The [cleanup step below](#different-cleanup-when-using-the-removed-block-for-a-module) shows how the `moved` and `removed` block can remove the `mongodbatlas_cloud_backup_schedule` from your Terraform state to avoid the [BACKUP_POLICIES_NOT_MEETING_BACKUP_COMPLIANCE_POLICY_REQUIREMENTS](../resource/README.md#4-cleanup-extra-steps-when-a-backup-compliance-policy-is-enabled) error when running `terraform destroy`.
```

Author: Thank you changed in db0d3201708ec0c9fc0fdc261d71ea1d9e2cc7f9

## Different cleanup when using the `removed` block for a module

**Note**: If you can use the `terraform state rm` command, follow the simpler steps in the [resource example readme](../resource/README.md#3-use-terraform-state-rm-to-remove-mongodbatlas_cloud_backup_schedule-from-the-state-to-avoid-the-delete-call-for-mongodbatlas_cloud_backup_schedule).

When using a module, modifying the module terraform files can be inaccessible:

> - The `module.source` can be pointing at an online repository, for example [terraform-mongodbatlas-atlas-basic](https://github.com/terraform-mongodbatlas-modules/terraform-mongodbatlas-atlas-basic)

Reviewer: in here you're just saying "when you don't have access to the terraform files", which I think it is not necessary.

Reviewer 2: If it might not be common knowledge for users, it should be included, but it should be stated more directly "If you can edit the TF files directly, follow these steps" or something similar.

Author: Thank you both, made it more concise

> - If you can modify the module terraform files, follow the steps in the [resource example readme](../resource/README.md#1-recommended-use-a-removed-block-to-avoid-the-delete-call-for-mongodbatlas_cloud_backup_schedule)

Reviewer: so you're saying: if you "module user" can directly change the source code of the module, change it as if you would do it directly using resources.

Is that right?

Author: Correct

> If you try to use the `removed` block without deleting the `from` resouce you get the error: `Removed Resource still exists error`:

Reviewer: I don't understand this part... 
> use the `removed` block without deleting the `from` resouce

Are you saying if you "module user" add this code (below) outside the module source code, you get an error?

and also note that misspell "resouce"

Author: Changed
